marconi/src/Marconi/Index/AddressDatum.hs

```diff
@@ -0,0 +1,517 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE GADTs              #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE QuasiQuotes        #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TupleSections      #-}
+
+-- | Module for indexing the datums for all addresses in the Cardano blockchain.
+
+-- This module will create the SQL tables:
+
+-- + table: address_datums
+--
+-- @
+--    |---------+------------+---------+------------|
+--    | address | datum_hash | slot_no | block_hash |
+--    |---------+------------+---------+------------|
+-- @
+--
+-- + table: datumhash_datum
+--
+-- @
+--    |------------+-------|
+--    | datum_hash | datum |
+--    |------------+-------|
+-- @
+
+-- To create these tables, we extract all transactions outputs from each transactions fetched with
+-- the chain-sync protocol of the local node.
+
+-- Here is a synopsis of the indexing algorithm.
+
+-- Each transaction output contains an address along with an optional datum hash or an optional inline
+-- datum (actual datum).
+-- In the inline datum scenario, we simply create an entry in the `address_datums` table with the hash
+-- of the datum, and add an entry in the `datumhash_datum` table.
+-- In the datum hash scenario, we create an entry in the `address_datums` table, but not in the
+-- `datumhash_datum` table, as we don't know the actual value of the datum and we can't infer it from
+-- the hash.
+-- In that last scenario, we can resolve in the datum hash in one of three ways (which we then simply
+-- add an entry in the `datumhash_datum` table):
+--
+--     * a different transaction output has an inline datum with that same hash
+--
+--     * a datum with that same hash has been found in the in transaction body
+--
+--     * a datum with that same hash was included in the witnesses for a Plutus spending script
+--     which was included in the transaction body
```

Comment:
Thank you for the detail, it helped a lot. 
I am a bit confused as to why we need this table as all the fields are already in the unspent_transaction table. I put more detail in the ticket, [PLT-662](https://input-output.atlassian.net/browse/PLT-662)

Reply:
Good point. Here’s what I see:

    Utxo indexer does not resolve the datum hashes as I expect them. The requirements of the Utxo indexer were simply that we store the datum hash and inline datum of the tx output

We could change the Utxo indexer to better resolve the datums. 

This seems definitely something we should try. My intuition with Address-Datum indexer is that queries should be faster because the table is smaller than the Utxo indexer. It would certainly make sense to compare them.

Reply:
Thnx @koslambrou . I am in favor of smaller tables as the UTXO table already had quite a few fileds.